contracts/BaseWrapper.sol

```diff
@@ -0,0 +1,162 @@
+    }
+
+    function _setRegistry(address _registry) internal {
+        // In case you want to override the registry instead of re-deploying
+        registry = RegistryAPI(_registry);
+    }
+
+    function bestVault() public virtual view returns (VaultAPI) {
+        return VaultAPI(registry.latestVault(address(token)));
+    }
+
+    function allVaults() public virtual view returns (VaultAPI[] memory) {
+        uint256 cache_length = _cachedVaults.length;
+        uint256 num_deployments = registry.nextDeployment(address(token));
+
+        // Use cached
+        if (cache_length == num_deployments) {
+            return _cachedVaults;
+        }
+
+        VaultAPI[] memory vaults = new VaultAPI[](num_deployments);
+
+        for (uint256 deployment_id = 0; deployment_id < cache_length; deployment_id++) {
+            vaults[deployment_id] = _cachedVaults[deployment_id];
+        }
+
+        for (uint256 deployment_id = cache_length; deployment_id < num_deployments; deployment_id++) {
+            vaults[deployment_id] = VaultAPI(registry.vaults(address(token), deployment_id));
+        }
+
+        return vaults;
+    }
+
+    function _updateVaultCache(VaultAPI[] memory vaults) internal {
+        if (vaults.length > 0) {
+            _cachedVaults = vaults;
+        }
+    }
+
+    function totalBalance(address account) public view returns (uint256 balance) {
+        VaultAPI[] memory vaults = allVaults();
+
+        for (uint256 id = 0; id < vaults.length; id++) {
+            balance = balance.add(vaults[id].balanceOf(account).mul(vaults[id].pricePerShare()).div(10**vaults[id].decimals()));
+        }
+    }
+
+    function totalAssets() public view returns (uint256 assets) {
+        VaultAPI[] memory vaults = allVaults();
+
+        for (uint256 id = 0; id < vaults.length; id++) {
+            assets = assets.add(vaults[id].totalAssets().mul(vaults[id].pricePerShare()).div(10**vaults[id].decimals()));
+        }
+    }
+
+    function _deposit(
+        address depositor,
+        address reciever,
+        uint256 amount,
+        bool pullFunds // If true, funds need to be pulled from `depositor` via `transferFrom`
+    ) internal returns (uint256) {
+        VaultAPI best = bestVault();
+
+        if (pullFunds) {
+            token.transferFrom(depositor, address(this), amount);
+        }
+
+        if (token.allowance(address(this), address(best)) < uint256(-1)) {
+            token.approve(address(best), uint256(-1)); // Vaults are trusted
+        }
+
+        // `receiver` now has shares of `best` (worth `amount` tokens) as balance
+        return
+            best
+                .deposit(amount, reciever)
+                .mul(10**3) // Add 3 extra decimals of precision
+                .mul(best.pricePerShare()) // Adjust by price of best
+                .div(10**best.decimals())
+                .div(10**3); // Add 3 extra decimals of precision
+    }
+
+    function _withdraw(
```

Comment:
thinking if there is value in having a method `_withdrawFromVault` has the id of the vault so user can signal to withdraw from a specific vault directly? it would add an additional id param and can check if id is the best vault to not allow it i guess? just thinking its an optimize withdraw option, of course it will fail if maxShares is higher than amount

Reply:
Related fix in 05b876b that allows limiting the migration amount. Should be a decent mitigation for the issue here..